File: packages/core/src/agent-overrides.ts
import { existsSync, readFileSync, statSync } from "node:fs";
import { extname, resolve } from "node:path";
import { parse as parseYaml } from "yaml";
import type { AgentSpec, WorkflowSpec } from "./types.js";

export type AgentOverrides = Record<string, AgentOverride>;

export interface AgentOverride {
  type?: "builtin" | "command";
  use?: string;
  model?: string;
  cwd?: string;
  env?: Record<string, unknown>;
  policy?: "read" | "full";
}

export interface AgentOverrideWarning {
  code: "AGENT_MODEL_CLEARED" | "INHERITED_AGENT_OVERRIDE_SKIPPED";
  agent: string;
  message: string;
}

export interface ApplyAgentOverridesOptions {
  inherited?: AgentOverrides;
}

export interface ApplyAgentOverridesResult {
  effectiveSpec: WorkflowSpec;
  agentOverrides: AgentOverrides;
  warnings: AgentOverrideWarning[];
}

const SUPPORTED_OVERRIDE_FIELDS = new Set(["type", "use", "model", "cwd", "env", "policy"]);
const SUPPORTED_FILE_EXTENSIONS = new Set([".json", ".yaml", ".yml"]);

export function parseAgentOverridesInput(value: string | undefined, cwd = process.cwd()): AgentOverrides | undefined {
  if (value === undefined) return undefined;

  const possiblePath = resolve(cwd, value);
  if (existsSync(possiblePath)) {
    const stat = statSync(possiblePath);
    if (stat.isDirectory()) {
      throw new Error("--agents must be a JSON/YAML file or inline JSON/YAML object, not a directory.");
    }
    const extension = extname(possiblePath).toLowerCase();
    if (!SUPPORTED_FILE_EXTENSIONS.has(extension)) {
      throw new Error("--agents file must use .json, .yaml, or .yml.");
    }
    const contents = readFileSync(possiblePath, "utf8");
    const parsed = extension === ".json" ? JSON.parse(contents) : parseYaml(contents);
    return validateAgentOverrides(parsed, "--agents");
  }

  if (looksLikePath(value)) {
    throw new Error(`--agents file not found: ${value}`);
  }

  return validateAgentOverrides(parseYaml(value), "--agents");
}

export function parseWorkflowSpecForOverrides(source: string): WorkflowSpec {
  const parsed = parseYaml(source);
  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    throw new Error("Workflow spec must be a YAML object.");
  }
  return parsed as WorkflowSpec;
}

export function serializeWorkflowSpecForOverrides(spec: WorkflowSpec): string {
  return JSON.stringify(spec);
}

export function emptyAgentOverrideResult(spec?: WorkflowSpec): ApplyAgentOverridesResult {
  return {
    effectiveSpec: spec ?? { version: 1, name: "", workflow: { steps: [] } },
    agentOverrides: {},
    warnings: []
  };
}

export function optionalSubmissionMetadata(result: ApplyAgentOverridesResult): {
  agentOverrides?: AgentOverrides;
  submissionWarnings?: AgentOverrideWarning[];
} {
  return {
    agentOverrides: Object.keys(result.agentOverrides).length > 0 ? result.agentOverrides : undefined,
    submissionWarnings: result.warnings.length > 0 ? result.warnings : undefined
  };
}

export function validateAgentOverrides(value: unknown, label = "Agent Overrides"): AgentOverrides {
  if (!isRecord(value)) {
    throw new Error(`${label} must resolve to an object.`);
  }
  const overrides: AgentOverrides = {};
  for (const [agentName, rawOverride] of Object.entries(value)) {
    if (!isRecord(rawOverride)) {
      throw new Error(`${label}.${agentName} must be an object.`);
    }
    const keys = Object.keys(rawOverride);
    if (keys.length === 0) {
      throw new Error(`${label}.${agentName} must not be empty.`);
    }
    for (const key of keys) {
      if (!SUPPORTED_OVERRIDE_FIELDS.has(key)) {
        throw new Error(`${label}.${agentName}.${key} is not supported.`);
      }
    }
    if ((rawOverride.type === undefined) !== (rawOverride.use === undefined)) {
      throw new Error(`${label}.${agentName} must specify type and use together.`);
    }
    if (rawOverride.type !== undefined && rawOverride.type !== "builtin" && rawOverride.type !== "command") {
      throw new Error(`${label}.${agentName}.type must be 'builtin' or 'command'.`);
    }
    if (rawOverride.use !== undefined && typeof rawOverride.use !== "string") {
      throw new Error(`${label}.${agentName}.use must be a string.`);
    }
    if (rawOverride.model === null) {
      throw new Error(`${label}.${agentName}.model must be a string; null model clearing is not supported.`);
    }
    if (rawOverride.model !== undefined && typeof rawOverride.model !== "string") {
      throw new Error(`${label}.${agentName}.model must be a string.`);
    }
    if ("cwd" in rawOverride) {
      if (typeof rawOverride.cwd !== "string" || rawOverride.cwd.length === 0) {
        throw new Error(`${label}.${agentName}.cwd must be a non-empty string.`);
      }
    }
    if (rawOverride.env !== undefined && !isRecord(rawOverride.env)) {
      throw new Error(`${label}.${agentName}.env must be an object.`);
    }
    if (rawOverride.policy !== undefined && rawOverride.policy !== "read" && rawOverride.policy !== "full") {
      throw new Error(`${label}.${agentName}.policy must be 'read' or 'full'.`);
    }

    const override: AgentOverride = {};
    if (rawOverride.type !== undefined) override.type = rawOverride.type;
    if (rawOverride.use !== undefined) override.use = rawOverride.use;
    if (rawOverride.model !== undefined) override.model = rawOverride.model;
    if ("cwd" in rawOverride) override.cwd = rawOverride.cwd as string;
    if (rawOverride.env !== undefined) override.env = { ...rawOverride.env };
    if (rawOverride.policy !== undefined) override.policy = rawOverride.policy;
    overrides[agentName] = override;
  }
  return overrides;
}

export function applyAgentOverrides(
  spec: WorkflowSpec,
  current: AgentOverrides | undefined,
  options: ApplyAgentOverridesOptions = {}
): ApplyAgentOverridesResult {
  const agents = isRecord(spec.agents) ? spec.agents : {};
  const effectiveSpec: WorkflowSpec = {
    ...spec,
    agents: cloneAgents(agents)
  };
  const finalOverrides: AgentOverrides = {};
  const warnings: AgentOverrideWarning[] = [];

  for (const [agentName, inheritedOverride] of Object.entries(options.inherited ?? {})) {
    const normalized = validateSingleOverride(agentName, inheritedOverride, "inherited Agent Overrides");
    if (!Object.prototype.hasOwnProperty.call(agents, agentName)) {
      warnings.push({
        code: "INHERITED_AGENT_OVERRIDE_SKIPPED",
        agent: agentName,
        message: `Inherited Agent Override for '${agentName}' was skipped because the repaired Workflow Spec does not declare that agent.`
      });
      continue;
    }
    applySingleOverride(effectiveSpec.agents!, finalOverrides, agentName, normalized, warnings);
  }

  for (const [agentName, currentOverride] of Object.entries(current ?? {})) {
    const normalized = validateSingleOverride(agentName, currentOverride, "Agent Overrides");
    if (!Object.prototype.hasOwnProperty.call(agents, agentName)) {
      throw new Error(`Agent Override '${agentName}' does not match a top-level agent declared by the Workflow Spec.`);
    }
    applySingleOverride(effectiveSpec.agents!, finalOverrides, agentName, normalized, warnings);
  }

  return {
    effectiveSpec,
    agentOverrides: finalOverrides,
    warnings
  };
}

function validateSingleOverride(agentName: string, value: unknown, label: string): AgentOverride {
  return validateAgentOverrides({ [agentName]: value }, label)[agentName]!;
}

function applySingleOverride(
  agents: Record<string, AgentSpec>,
  finalOverrides: AgentOverrides,
  agentName: string,
  override: AgentOverride,
  warnings: AgentOverrideWarning[]
): void {
  const currentAgent = agents[agentName] ?? {};
  const currentOverride = finalOverrides[agentName] ?? {};
  const identityChanged = override.type !== undefined || override.use !== undefined;
  const mergedAgent: AgentSpec = { ...currentAgent };
  const mergedOverride: AgentOverride = { ...currentOverride };

  if (identityChanged) {
    mergedAgent.type = override.type;
    mergedAgent.use = override.use;
    mergedOverride.type = override.type;
    mergedOverride.use = override.use;
    // Model is identity-tied (different adapters use different models), so
    // clearing it on identity change prevents sending a stale model to a new
    // adapter. Policy is orthogonal to agent identity — a read-only reviewer
    // should stay read-only regardless of which adapter backs it — so policy
    // is preserved rather than cleared.
    if (override.model === undefined && mergedAgent.model !== undefined) {
      delete mergedAgent.model;
      delete mergedOverride.model;
      warnings.push({
        code: "AGENT_MODEL_CLEARED",
        agent: agentName,
        message: `Agent Override for '${agentName}' changed type/use and cleared the inherited model.`
      });
    }
  }
  if (override.model !== undefined) {
    mergedAgent.model = override.model;
    mergedOverride.model = override.model;
  }
  if ("cwd" in override) {
    mergedAgent.cwd = override.cwd;
    mergedOverride.cwd = override.cwd;
  }
  if (override.env !== undefined) {
    mergedAgent.env = { ...(isRecord(mergedAgent.env) ? mergedAgent.env : {}), ...override.env };
    mergedOverride.env = { ...(isRecord(mergedOverride.env) ? mergedOverride.env : {}), ...override.env };
  }
  if (override.policy !== undefined) {
    mergedAgent.policy = override.policy;
    mergedOverride.policy = override.policy;
  }

  agents[agentName] = mergedAgent;
  finalOverrides[agentName] = mergedOverride;
}

function cloneAgents(agents: Record<string, AgentSpec>): Record<string, AgentSpec> {
  return Object.fromEntries(
    Object.entries(agents).map(([name, agent]) => [
      name,
      {
        ...agent,
        env: isRecord(agent.env) ? { ...agent.env } : agent.env
      }
    ])
  );
}

function looksLikePath(value: string): boolean {
  const trimmed = value.trim();
  if (trimmed.startsWith("{") || trimmed.includes("\n") || /^[A-Za-z0-9_.-]+\s*:/.test(trimmed)) {
    return false;
  }
  return trimmed.startsWith(".")
    || trimmed.startsWith("/")
    || trimmed.startsWith("~")
    || trimmed.includes("/")
    || SUPPORTED_FILE_EXTENSIONS.has(extname(trimmed).toLowerCase());
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: crates/acpus-core/src/agent_overrides.rs
use crate::{AcpusIr, AgentPolicy, AgentSpec, AgentType, IrNode, IrNodeKind};
use serde::{Deserialize, Serialize};
use serde_json::{Value, to_value};
use std::collections::{BTreeMap, BTreeSet};

pub type AgentOverrides = BTreeMap<String, AgentOverride>;

#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
pub struct AgentOverride {
    #[serde(default, rename = "type", skip_serializing_if = "Option::is_none")]
    pub agent_type: Option<AgentType>,
    #[serde(default, rename = "use", skip_serializing_if = "Option::is_none")]
    pub use_: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cwd: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub env: Option<BTreeMap<String, Value>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub policy: Option<AgentPolicy>,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct AgentOverrideWarning {
    pub code: String,
    pub agent: String,
    pub message: String,
}

#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
pub struct ApplyAgentOverridesResult {
    #[serde(rename = "agentOverrides")]
    pub agent_overrides: AgentOverrides,
    #[serde(rename = "submissionWarnings")]
    pub warnings: Vec<AgentOverrideWarning>,
}

pub fn validate_agent_overrides(value: &Value, label: &str) -> anyhow::Result<AgentOverrides> {
    let Some(root) = value.as_object() else {
        anyhow::bail!("{label} must resolve to an object.");
    };
    let supported = BTreeSet::from(["type", "use", "model", "cwd", "env", "policy"]);
    let mut out = AgentOverrides::new();
    for (agent, raw) in root {
        let Some(map) = raw.as_object() else {
            anyhow::bail!("{label}.{agent} must be an object.");
        };
        anyhow::ensure!(!map.is_empty(), "{label}.{agent} must not be empty.");
        for key in map.keys() {
            anyhow::ensure!(
                supported.contains(key.as_str()),
                "{label}.{agent}.{key} is not supported."
            );
        }
        anyhow::ensure!(
            map.contains_key("type") == map.contains_key("use"),
            "{label}.{agent} must specify type and use together."
        );
        let agent_type = match map.get("type") {
            Some(Value::String(value)) if value == "builtin" => Some(AgentType::Builtin),
            Some(Value::String(value)) if value == "command" => Some(AgentType::Command),
            Some(_) => anyhow::bail!("{label}.{agent}.type must be 'builtin' or 'command'."),
            None => None,
        };
        let use_ = match map.get("use") {
            Some(Value::String(value)) => Some(value.clone()),
            Some(_) => anyhow::bail!("{label}.{agent}.use must be a string."),
            None => None,
        };
        let model = match map.get("model") {
            Some(Value::String(value)) => Some(value.clone()),
            Some(Value::Null) => anyhow::bail!(
                "{label}.{agent}.model must be a string; null model clearing is not supported."
            ),
            Some(_) => anyhow::bail!("{label}.{agent}.model must be a string."),
            None => None,
        };
        let cwd = match map.get("cwd") {
            Some(Value::String(value)) if !value.is_empty() => Some(value.clone()),
            Some(_) => anyhow::bail!("{label}.{agent}.cwd must be a non-empty string."),
            None => None,
        };
        let env = match map.get("env") {
            Some(Value::Object(value)) => Some(value.clone().into_iter().collect()),
            Some(_) => anyhow::bail!("{label}.{agent}.env must be an object."),
            None => None,
        };
        let policy = match map.get("policy") {
            Some(Value::String(value)) if value == "read" => Some(AgentPolicy::Read),
            Some(Value::String(value)) if value == "full" => Some(AgentPolicy::Full),
            Some(_) => anyhow::bail!("{label}.{agent}.policy must be 'read' or 'full'."),
            None => None,
        };
        out.insert(
            agent.clone(),
            AgentOverride {
                agent_type,
                use_,
                model,
                cwd,
                env,
                policy,
            },
        );
    }
    Ok(out)
}

pub fn apply_agent_overrides(
    ir: &mut AcpusIr,
    current: Option<&AgentOverrides>,
    inherited: Option<&AgentOverrides>,
) -> anyhow::Result<ApplyAgentOverridesResult> {
    let mut final_overrides = AgentOverrides::new();
    let mut warnings = Vec::new();
    for (agent, override_) in inherited.into_iter().flat_map(BTreeMap::iter) {
        if !ir.agents.contains_key(agent) {
            warnings.push(AgentOverrideWarning {
                code: "INHERITED_AGENT_OVERRIDE_SKIPPED".to_string(),
                agent: agent.clone(),
                message: format!(
                    "Inherited Agent Override for '{agent}' was skipped because the repaired Workflow Spec does not declare that agent."
                ),
            });
            continue;
        }
        apply_single_override(
            &mut ir.agents,
            &mut final_overrides,
            agent,
            override_,
            &mut warnings,
        );
    }
    for (agent, override_) in current.into_iter().flat_map(BTreeMap::iter) {
        anyhow::ensure!(
            ir.agents.contains_key(agent),
            "Agent Override '{agent}' does not match a top-level agent declared by the Workflow Spec."
        );
        apply_single_override(
            &mut ir.agents,
            &mut final_overrides,
            agent,
            override_,
            &mut warnings,
        );
    }
    refresh_agent_metadata(ir);
    Ok(ApplyAgentOverridesResult {
        agent_overrides: final_overrides,
        warnings,
    })
}

pub(crate) fn refresh_agent_metadata(ir: &mut AcpusIr) {
    let agents = ir.agents.clone();
    refresh_node_agent_metadata(&mut ir.root, &agents);
}

fn refresh_node_agent_metadata(node: &mut IrNode, agents: &BTreeMap<String, AgentSpec>) {
    if node.kind == IrNodeKind::RunAgent
        && let Some(metadata) = node.metadata.as_object_mut()
    {
        if let Some(agent) = metadata
            .get("use")
            .and_then(Value::as_str)
            .and_then(|name| agents.get(name))
            .and_then(|agent| to_value(agent).ok())
        {
            metadata.insert("agent".to_string(), agent);
        } else {
            metadata.remove("agent");
        }
    }
    for child in &mut node.children {
        refresh_node_agent_metadata(child, agents);
    }
    for branch in &mut node.branches {
        refresh_node_agent_metadata(&mut branch.child, agents);
    }
}

fn apply_single_override(
    agents: &mut BTreeMap<String, AgentSpec>,
    final_overrides: &mut AgentOverrides,
    agent: &str,
    override_: &AgentOverride,
    warnings: &mut Vec<AgentOverrideWarning>,
) {
    let spec = agents.entry(agent.to_string()).or_default();
    let final_override = final_overrides.entry(agent.to_string()).or_default();
    let identity_changed = override_.agent_type.is_some() || override_.use_.is_some();
    if identity_changed {
        spec.agent_type = override_.agent_type.clone().unwrap_or_default();
        spec.use_ = override_.use_.clone();
        final_override.agent_type = override_.agent_type.clone();
        final_override.use_ = override_.use_.clone();
        if override_.model.is_none() && spec.model.take().is_some() {
            final_override.model = None;
            warnings.push(AgentOverrideWarning {
                code: "AGENT_MODEL_CLEARED".to_string(),
                agent: agent.to_string(),
                message: format!(
                    "Agent Override for '{agent}' changed type/use and cleared the inherited model."
                ),
            });
        }
    }
    if let Some(model) = &override_.model {
        spec.model = Some(model.clone());
        final_override.model = Some(model.clone());
    }
    if let Some(cwd) = &override_.cwd {
        spec.cwd = Some(Value::String(cwd.clone()));
        final_override.cwd = Some(cwd.clone());
    }
    if let Some(env) = &override_.env {
        spec.env.extend(env.clone());
        final_override
            .env
            .get_or_insert_with(BTreeMap::new)
            .extend(env.clone());
    }
    if let Some(policy) = &override_.policy {
        spec.policy = policy.clone();
        final_override.policy = Some(policy.clone());
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{CompileOptions, compile_workflow};
    use serde_json::json;

    #[test]
    fn validates_override_shape() {
        assert!(validate_agent_overrides(&json!([]), "--agents").is_err());
        assert!(
            validate_agent_overrides(&json!({"implementer": {"policy": "read"}}), "--agents")
                .is_ok()
        );
        assert!(
            validate_agent_overrides(&json!({"implementer": {"type": "builtin"}}), "--agents")
                .is_err()
        );
    }

    #[test]
    fn applies_overrides_and_warns_when_identity_clears_model() {
        let mut ir = compile_workflow(
            r#"
version: 1
name: t
agents:
  implementer:
    type: builtin
    use: codex
    model: gpt-5
workflow:
  steps:
    - id: a
      run: agent
      use: implementer
      prompt: hi
"#,
            CompileOptions::default(),
        )
        .ir
        .unwrap();
        let overrides = validate_agent_overrides(
            &json!({"implementer": {"type": "command", "use": "agent.sh"}}),
            "--agents",
        )
        .unwrap();

        let result = apply_agent_overrides(&mut ir, Some(&overrides), None).unwrap();

        assert_eq!(ir.agents["implementer"].use_, Some("agent.sh".to_string()));
        assert_eq!(ir.agents["implementer"].model, None);
        assert_eq!(
            ir.root.children[0].metadata["agent"]["use"],
            Value::String("agent.sh".to_string())
        );
        assert!(ir.root.children[0].metadata["agent"].get("model").is_none());
        assert_eq!(result.warnings[0].code, "AGENT_MODEL_CLEARED");
    }
}
